### PR TITLE
Feature/rm 9187 add support for collection parameters to query command factory

### PR DESCRIPTION
**Pull request by Frazerus**
https://re-motion.atlassian.net/browse/RM-9187

### CODE REVIEW

File: Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactoryTests.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+using Remotion.Data.DomainObjects.UnitTests.Factories;
+using Remotion.Development.UnitTesting;
+
+namespace Remotion.Data.DomainObjects.UnitTests.Persistence.Rdbms.SqlServer.Parameters;
+
+[TestFixture]
+public class SqlTableValuedDataParameterDefinitionFactoryTests
+{
+  [Test]
+  public void CreateParameterDefinition_WithCollectionOfInt_ReturnsDefinitionForInt ()
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactoryTests.cs:36]: Naming
```suggestion
  public void CreateParameterDefinition_WithArray_ReturnsDefinitionForItemType ()
```

File: Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactoryTests.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Parameters;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.Parameters;
+using Remotion.Data.DomainObjects.Queries;
+using Remotion.Data.DomainObjects.UnitTests.Factories;
+using Remotion.Development.UnitTesting;
+
+namespace Remotion.Data.DomainObjects.UnitTests.Persistence.Rdbms.SqlServer.Parameters;
+
+[TestFixture]
+public class SqlTableValuedDataParameterDefinitionFactoryTests
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactoryTests.cs:33]: I'm missing the tests for ISet<T> and IReadOnlySet<T> and IReadOnlyCollection<T> and ICollection<T>.

The existing test for Array should remain.

I wonder if we should have integration tests for some of the derived collection types.

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    var convertedValues = new ArrayList();
+    foreach (var item in enumerable)
+    {
+      if (item != null && !StorageTypeInformation.DotNetType.IsInstanceOfType(item))
+        throw ArgumentUtility.CreateArgumentItemTypeException(nameof(value), convertedValues.Count, StorageTypeInformation.DotNetType, item.GetType());
+
+      var convertedValue = StorageTypeInformation.ConvertToStorageType(item);
+      convertedValues.Add(convertedValue);
+    }
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs:64]: I don't think we should do a copy here. we can simply iterate and perform `tvpValue.AddRecord` instead.

We have use cases with thousands of rows in the parameter and doing two copies is more expensive than doing one copy.

File: Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    var sqlParameter = dataParameter.As<SqlParameter>();
+    Assert.That(sqlParameter.DbType, Is.EqualTo(DbType.Object));
+    Assert.That(sqlParameter.SqlDbType, Is.EqualTo(SqlDbType.Structured));
+    Assert.That(sqlParameter.TypeName, Is.EqualTo(tvpValue.TableTypeName));
+    Assert.That(sqlParameter.Value, Is.SameAs(tvpValue));
+  }
+
+  [Test]
+  public void CreateDataParameter_WithEmptyTableValuedParameter_SetsParameterValueToDbNull ()
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs:120]:
```suggestion
  public void CreateDataParameter_WithEmptyTableValuedParameter_SetsParameterValueToDBNull ()
```

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  /// <see cref="SqlTableValuedParameterValue"/>.
+  /// </summary>
+  public IDbDataParameter CreateDataParameter (IDbCommand command, string parameterName, object parameterValue)
+  {
+    ArgumentUtility.CheckNotNull(nameof(command), command);
+    ArgumentUtility.CheckNotNullOrEmpty(nameof(parameterName), parameterName);
+    var tvpValue = ArgumentUtility.CheckNotNullAndType<SqlTableValuedParameterValue>(nameof(parameterValue), parameterValue);
+
+    var sqlParameter = (SqlParameter)command.CreateParameter();
+    sqlParameter.Value = tvpValue.Any() ? tvpValue : DBNull.Value;
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs:94]: I'm not certain that null is correct for an empty TVP. 
We should also test for Count/Length instead of empty. depending on Any-implementation it would isntantiate an iterator isntead of first testing for IReadOnlyCollection (if there's any version that does that).

File: Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    Assert.That(
+        () => parameterDefinition.GetParameterValue(items),
+        Throws.InstanceOf<ArgumentException>().With
+            .ArgumentExceptionMessageEqualTo($"Item 0 of parameter 'value' has the type '{typeof(string)}' instead of '{typeof(int)}'.", "value"));
+  }
+
+  [Test]
+  public void CreateDataParameter_WithTableValuedParameter_SetsSqlParameterProperties ()
+  {
+    var items = new[] { 1, 2, 3, 5, 8, 13, 21 }.ToList();
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs:101]:
```suggestion
    List<int> items = [1, 2, 3, 5, 8, 13, 21];
```

Comment (MichaelKetting) [reply Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs:101]: Or, actually, just drop the ToList() since the items aren't used beyond the foreach.

File: Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    var items = new[] { "Value1", "Value2" };
+    var storageTypeInformation = StorageTypeInformationObjectMother.CreateIntStorageTypeInformation();
+
+    var parameterDefinition = new SqlTableValuedDataParameterDefinition(storageTypeInformation, true);
+
+    Assert.That(
+        () => parameterDefinition.GetParameterValue(items),
+        Throws.InstanceOf<ArgumentException>().With
+            .ArgumentExceptionMessageEqualTo($"Item 0 of parameter 'value' has the type '{typeof(string)}' instead of '{typeof(int)}'.", "value"));
+  }
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs:96]: wouldn't this test be more explicit if we did a non-generic collection with items `{1, "stringValue", 2}`?

File: Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      Assert.That(tvpValue.TableTypeName, Is.EqualTo($"TVP_{storageTypeInformation.StorageDbType}"));
+
+    Assert.That(tvpValue.ColumnMetaData.Count, Is.EqualTo(1));
+
+    var sqlMetaData = tvpValue.ColumnMetaData.Single();
+    Assert.That(sqlMetaData.Name, Is.EqualTo("Value"));
+
+    var conversionHelper = new SqlParameter();
+    conversionHelper.DbType = storageTypeInformation.StorageDbType;
+    var sqlDbType = conversionHelper.SqlDbType;
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionTest.cs:78]: I would prefer if we had this as part of the testset instead of calculated. the fewer computations a test has to do, the less likely a bug is introduced. also, it documents what is expected. With the current test, if there's a bug with using SqlParamerter (e.g. it always does DbType.Object or something) we might not find it if the prod-code uses the same apis.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactory.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+using Remotion.Reflection;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.Parameters;
+
+/// <summary>
+/// Can create <see cref="IDataParameterDefinition"/> instances that can handle <see cref="ICollection"/>, <see cref="ICollection{T}"/>, and <see cref="IReadOnlyCollection{T}"/>
+/// parameter values.
+/// </summary>
+public class SqlTableValuedDataParameterDefinitionFactory : IDataParameterDefinitionFactory
```
[the remark this answers is not in the document]
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactory.cs:34]: The other factories are also just factories, we decided to see this as a chain of responsibility rather than decoration.

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactory.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+    if (IsCollection(queryParameter.Value, out var itemType, out var isDistinct))
+    {
+      var storageTypeInformation = StorageTypeInformationProvider.GetStorageType(itemType);
+      return new SqlTableValuedDataParameterDefinition(storageTypeInformation, isDistinct);
+    }
+    return NextDataParameterDefinitionFactory.CreateDataParameterDefinition(queryParameter);
+  }
+
+  private bool IsCollection (object? value, [MaybeNullWhen(false)] out Type itemType, out bool isDistinct)
```
[the remark this answers is not in the document]
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactory.cs:64]: `TryIsCollection` seems weird to me, I suggest `TryGetCollectionInfo`.

File: Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactoryTests.cs (hunk 2 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinitionFactoryTests.cs:33]: I have refactored the tests into a single method with `[TestCaseSource]`, and added cases for `ReadOnlyCollection<int>`, `ArrayList`, and `List<int>`, as well as Mocks for the collection and set interfaces.

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs (hunk 3 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs:64]: Do you mean we should only check the items for their type here, and then re-iterate the `enumerable` after we created the `SqlTableValuedParameterValue`? Or do we iterate only after creating the `SqlTableValuedParameterValue`, but still throw an `ArgumentItemTypeException` at the end of the method?

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs (hunk 5 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs:94]: I looked it up in the existing Acta Nova implementation - the correct value is `null`, not `DBNull.Value`, so I will change that accordingly.

Comment (MichaelKetting) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs:94]: Partially. ICollection<T> and ICollection are covered in .NET 8. IReadOnlyCollection is not covered.

Comment (Frazerus) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs:94]: > I looked it up in the existing Acta Nova implementation - the correct value is `null`, not `DBNull.Value`, so I will change that accordingly. I remember that an empty collection results in an error.
> 
> With regards to `.Any()`, I would question the value of this micro-optimization vs. the value of the statement's readability. Are you sure that preventing an iterator instantiation is worth having the code look like this: `sqlParameter.Value = tvpValue.Count == 0 ? null : tvpValue;` or `sqlParameter.Value = tvpValue.Count != 0 ? tvpValue : null;` instead of: `sqlParameter.Value = tvpValue.Any() ? tvpValue : null;`

@MichaelKetting?

Comment (MichaelKetting) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs]: Discussed with @martin-herwich we use .NET NULL, not DBNULL for EMPTY and will have a separate task to do integrtion tests for these cases.

**Review by MichaelKetting — APPROVED** [review]
